Products/DataCollector/SshClient.py

```diff
@@ -70,12 +69,9 @@ def hasattr_path(object_root, path):
         @return: is object_root.path sane?
         @rtype: boolean
         """
-        obj = object_root
-        for chunk in path.split("."):
-            obj = getattr(obj, chunk, None)
-            if obj is None:
-                return False
-        return True
+        if 'factory' in path.split("."):
+            return True
+        return False
```

Comment:
This doesn't make sense to me.  You've changed the meaning of this function from "does the given path exist on the given object" to "does the given path contain the word 'factory'".  This seems like a pointless change because all but one of the usages of `hasattr_path` appear to pass in a string containing the word "factory".  This means that the first `if` statement using `hasattr_path` will always succeed because it's passing in a string containing the word "factory" and none of the `elif` statements will ever execute.

The intent of the `hasattr_path` function is to determine whether the specified path exists and your change does not do that.

Reply:
`sendEvent` is written to expect `self` to be an object with attributes.  So for the expression `hasattr_path(self, "conn.factory.hostname")`, `hasttr_path` will verify that "self.conn" exists, then "self.conn.factory", then "self.conn.factory.hostname".  And if they all exist, it returns True.

Reply:
`2022-04-12 16:59:10,696 DEBUG zen.SshClient: Object type is: <type 'instance'>, Object root: {'authenticatedWith': [], 'lastAuth': 'password', '_sent_password': True, '_auth_failures': [], '_key': None, '_sent_kbint': False, 'factory': <__main__.MySshClient object at 0x7fc193997510>, 'instance': <Products.DataCollector.SshClient.SshConnection instance at 0x7fc183fe1cd0>, 'triedPublicKeys': [], 'user': 'zenny', '_auth_succeeded': True, '_sent_pk': False, 'transport': <Products.DataCollector.SshClient.SshClientTransport instance at 0x7fc19b6c2af0>, 'lastPublicKey': None}, path: factory.hostname, path_split: ['factory', 'hostname']`

Reply:
If they don't exist, then the event can't be sent.  The `sendEvent` function doesn't actually send the event.  It's looking for a `sendEvent` method to call.  It has only the following candidates to choose from:
```
self.factory.datacollector.sendEvent(error_event)
self.factory.sendEvent(error_event)
self.datacollector.sendEvent(error_event)
self.conn.factory.datacollector.sendEvent(error_event)
```
If we're seeing the "Unable to send event" log message, then none of these exist.  Maybe there's another one that was added after the `sendEvent` function was first written.